FABulous.py

```diff
@@ -0,0 +1,743 @@
+    def do_gen_bitstream(self, args):
+        logger.info(f"Generating Bitstream for design {self.top}")
+        runCmd = ["python3", f"./nextpnr/fabulous/fab_arch/bit_gen.py",
+                  "-genBitstream",
+                  f"./{self.projectDir}/{self.top}.fasm",
+                  f"{self.projectDir}/.FABulous/meta_data.txt",
+                  f"./{self.projectDir}/{self.top}.bin"]
+
+        sp.run(runCmd, check=True)
+        logger.info("Bitstream generated")
+
+
+if __name__ == "__main__":
+    if sys.version_info < (3, 8, 10):
+        print("Need Python 3.8.10 or above to run FABulous")
+        exit(-1)
+    parser = argparse.ArgumentParser(
+        description='The command line interface for FABulous')
+
+    parser.add_argument(
+        "project_dir", help="The directory to the project folder")
+
+    parser.add_argument('-c', '--createProject',
+                        default=False,
+                        action='store_true',
+                        help='Create a new project')
+
+    parser.add_argument('-csv',
+                        default="",
+                        nargs=1,
+                        help="Log all the output from the terminal")
+
+    parser.add_argument('-s', '--script',
+                        default=False,
+                        nargs=1,
+                        help="Run FABulous with a FABulous script")
+
+    parser.add_argument('-log',
+                        default=False,
+                        nargs='?',
+                        const="FABulous.log",
+                        help="Log all the output from the terminal")
+
+    parser.add_argument('-w', '--writer',
+                        default="verilog",
+                        nargs=1,
+                        choices=['verilog', 'vhdl'],
+                        help="Set the type of HDL code generated by the tool. Currently support Verilog and VHDL (Default using Verilog)")
+
+    parser.add_argument('-md', '--metaDataDir',
+                        default=".FABulous",
+                        nargs=1,
+                        help="Set the output directory for the meta data files eg. pip.txt, bel.txt")
+
+    args = parser.parse_args()
+
+    args.top = args.project_dir.split("/")[-1]
+    metaDataDir = ".FABulous"
+
+    if args.createProject:
+        create_project(args.project_dir, args.writer)
+        exit(0)
+
+    if not os.path.exists(f"{args.project_dir}/.FABulous"):
+        print("The directory provided is not a FABulous project as it does not have a .FABulous folder")
+        exit(-1)
+    else:
+        writer = VerilogWriter("")
+        if args.writer == "vhdl":
+            writer = VHDLWriter("")
+        if args.writer == "verilog":
+            writer = VerilogWriter("")
+
+        os.chdir(args.project_dir)
+
+        fabShell = FABulousShell(
+            FABulous(writer, fabricCSV=args.csv), args.project_dir)
+
+        if args.script:
+            fabShell.cmdqueue.extend(args.script.splitlines())
```

Comment:
this doesn't seem to be working - I think you need to either open the file explicitly or tell argparse that it's a file argument